rfd/0073-kube-secret-storage.md

```diff
@@ -0,0 +1,501 @@
+                "tls_cert": "{tls_cert_content}",
+                "tls_ca_certs": ["tls_ca_certs"],
+                "ssh_ca_certs": ["ssh_ca_certs"]
+            }
+            }
+    
+kind: Secret
+metadata:
+  name: {.Release.Name}-identity
+  namespace: {.Release.Namespace}
+```
+
+Where:
+
+- `ssh_cert` is a PEM encoded SSH host cert.
+- `key` is a PEM encoded private key.
+- `tls_cert` is a PEM encoded x509 client certificate.
+- `tls_ca_certs` is a list of PEM encoded x509 certificate of the certificate authority of the cluster.
+- `ssh_ca_certs` is a list of SSH certificate authorities encoded in the authorized_keys format.
+- `TELEPORT_REPLICA_NAME` is the teleport agent replica name. Constant when using Deployments, `TELEPORT_REPLICA_NAME={{ .Release.Name}}` or dynamic when using StatefullSets, `TELEPORT_REPLICA_NAME=metadata.name`.
+
+#### RBAC Changes
+
+The Teleport Kube Agent service account must be able to read and edit secrets within the namespace that is running, therefore, one must create a new namespace role and attach it to the Kube Agent service account with the following content:
+
+```yaml
+- apiGroups: [""]
+  # objects is "secrets"
+  resources: ["secrets"]
+  verbs: ["create", "get", "update","watch", "list"]
+```
+
+This allows the Kube Agent to read, update and list secrets using the credentials that Kubernetes mounts for each pod running in the cluster.
+
+### Teleport Changes
+
+#### Kube Secret as storage backend
+
+If secret storage is enabled, the Teleport Kube agent initializes with Kubernetes secret [backend storage](https://goteleport.com/docs/setup/reference/backends/). The backend storage availability for Teleport will be the following:
+
+| Data type | Description | Supported storage backends |
+|---|---|---|
+| core cluster state | Cluster configuration (e.g. users, roles, auth connectors) and identity (e.g. certificate authorities, registered nodes, trusted clusters). | Local directory (SQLite), etcd, AWS DynamoDB, GCP Firestore, self-hosted PostgreSQL/CockroachDB (Preview) |
+| audit events | JSON-encoded events from the audit log (e.g. user logins, RBAC changes) | Local directory, AWS DynamoDB, GCP Firestore |
+| session recordings | Raw terminal recordings of interactive user sessions | Local directory, AWS S3 (and any S3-compatible product), GCP Cloud Storage |
+| teleport instance state | ID and credentials of a non-auth teleport instance (e.g. node, proxy, kube) | Local directory, Kube Secret (only available for role=kube) |
+
+The storage backend will be responsible for managing the Kubernetes secret, i.e. reading and updating its contents, in order to create a transparent storage backend.
+
+The configuration of the Kubernetes' Secret storage backend is done by adding the `storage` section to the teleport config.
+
+```yaml
+teleport:
+...
+  storage:
+    # Type of backend: either "local" or "kubernetes_secret", by default "local"
+    type: kubernetes_secret
```

Comment:
I'm not sure why storage backend changes are needed. We aren't going to support kubernetes as backend for everything, just for identities.

I think that it just should be the default mode of operation - when running in Kubernetes, always save identities in secrets. No extra configuration needed.

Reply:
Is it possible that the user wants to run in a mode where the identity is not stored in secrets (secrets in kube are not really secrets)? 
for example, for backup purposes, it might be something that some users prefer to use instead of Kube secrets. But if that's not the case it will simplify a lot

Reply:
Yep, exactly, I don't imagine a scenario where a user would want to store identity locally when running in Kubernetes.  So I would remove the backend configuration altogether.

Reply:
Not locally but if he wants to use Statefulset with PV storage in order to avoid losing events like agent `state` that are currently written to `sqlite.db`,  does it make sense to store it in secrets as well?